### PR TITLE
BufferGeometry computeBoundingSphere support for itemSize = 2

### PR DESCRIPTION
**Pull request by rock7-dan**
PR for Issue #6288 - added Box2.setFromArray for consistency with Box3

### CODE REVIEW

**Comment by mrdoob** [comment]
I suspect someone will want to support `itemSize = 4`...

**Comment by rock7-dan** [comment]
I'll amend ;-)

**Comment by rock7-dan** [comment]
Updated

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -604,17 +599,34 @@ THREE.BufferGeometry.prototype = {
...
 			if ( positions ) {
 
-				var center = this.boundingSphere.center;
+				var vector, box, center;
+
+				var itemSize = positions.itemSize;
+
+				if ( itemSize >= 3 ) {
+
+					vector = new THREE.Vector3();
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: I would avoid removing closures.

File: src/core/BufferGeometry.js [hunk 2]
```diff
@@ -604,17 +599,34 @@ THREE.BufferGeometry.prototype = {
...
+					box = new THREE.Box3();
+					center = this.boundingSphere.center;
+
+				} else {
+
+					vector = new THREE.Vector2();
+					box = new THREE.Box2();
+					center = new THREE.Vector2( this.boundingSphere.center.x, this.boundingSphere.center.y );
+
```
Comment (WestLangley) [inline src/core/BufferGeometry.js]: I do not think we want to go down this path.

Comment (mrdoob) [inline src/core/BufferGeometry.js]: Yeah... Me neither.

**Comment by mrdoob** [comment]
I think a better solution would be to pass `itemSize` to `Box3`'s  `setFromArray()`.

**Comment by rock7-dan** [comment]
Thanks for the quick feedback guys. 

Using `Vector3.fromArray` as-is will too cause problems - specifically if the size of the array is not divisible by 3 (i.e. undefined z value on the last position, which will lead to NaN when computing the radius). So I'd propose something like this to mitigate the NaN.

```
fromArray: function ( array, offset ) {

		if ( offset === undefined ) offset = 0;

		this.x = array[ offset ];
		this.y = array[ offset + 1 ];
		this.z = array[ offset + 2 ] || 0;
		
		return this;

	},
```

Likewise, with `Box3.setFromArray` even with a configurable `itemSize` there's a possibility for an undefined z value on the last position:

```
setFromArray: function ( array, itemSize ) {

		itemSize = itemSize || 3;

		this.makeEmpty();

		var minX = + Infinity;
		var minY = + Infinity;
		var minZ = + Infinity;

		var maxX = - Infinity;
		var maxY = - Infinity;
		var maxZ = - Infinity;

		for ( var i = 0, il = array.length; i < il; i += itemSize ) {

			var x = array[ i ];
			var y = array[ i + 1 ];
			var z = array[ i + 2 ] || minZ;
			
			if ( x < minX ) minX = x;
			if ( y < minY ) minY = y;
			if ( z < minZ ) minZ = z;

			if ( x > maxX ) maxX = x;
			if ( y > maxY ) maxY = y;
			if ( z > maxZ ) maxZ = z;

		}

		this.min.set( minX, minY, minZ );
		this.max.set( maxX, maxY, maxZ );

	}
```

Does this sound like a better approach?

**Comment by mrdoob** [comment]
It does sound better. But I'm still not sure...

**Comment by WestLangley** [comment]
How does it make sense to compute a bouning sphere in 3D when the data has only two dimensions?

**Comment by mrdoob** [comment]
The issue is that, for memory reasons, one could generate a geometry using a `position` TypeArray with only `x,y,x,y,...` values... One could also generate a geometry using `x,y,z,w,x,y,z,w,...` and use the `w` on their custom vertex shader.

Currently the library assumes `position` is `x,y,z,x,y,z,...` to compute the boundingBox/Sphere. How should it handle these cases? Should the user compute the boundingBox/Sphere themselves?

**Comment by WestLangley** [comment]
@mrdoob 

See https://github.com/mrdoob/three.js/issues/6288#issuecomment-204772958

`BufferGeometry.applyMatrix()` for example.

A solution for users with 2D data is to set the z-component to zero.

**Comment by mrdoob** [comment]
Yes, I understand. But, apart from the code that computes the boundingSphere/Box? Is there any other code that relies on `position` having `itemSize = 3`? If not, we can just say that for such cases the user will have to compute the boundingSphere/Box themselves...

**Comment by WestLangley** [comment]
>But, apart from the code that computes the boundingSphere/Box? Is there any other code that relies on position having itemSize = 3?

    .applyMatrix() and all the methods that call it
    .fromDirectGeometry()
    .computeVertexNormals()

**Comment by rock7-dan** [comment]
`itemSize != 3` has also caught me out within the `raycast` functions too, i.e:
https://github.com/mrdoob/three.js/blob/master/src/objects/Points.js#L78-L117

I fully appreciate the three in three.js - however, It seems a little superfluous to have to shoehorn in a third more geometry when working with 2d geometries; especially when WebGL renders it correctly as-is.

**Comment by mrdoob** [comment]
I hear you but I wonder what's the performance penalty of your patch though...

Also, do you need this for memory reasons? What's your use case?

**Comment by rock7-dan** [comment]
Yeah I am working with high resolution world maps and as such trying to keep usage to a minimum.

Let me put together a jsPerf for a comparison of the two approaches.

**Comment by Mugen87** [comment]
Closing, see https://github.com/mrdoob/three.js/issues/6288#issuecomment-226099732

**Pull request closed without merging**